### PR TITLE
Data validator

### PR DESCRIPTION
**Pull request by afonsobspinto**
closes (1st iteration of) https://metacell.atlassian.net/browse/YALE-6

- Removes old parsers
- Adds new parsers, following the rules defined [here](https://docs.google.com/presentation/d/1Zr5cRkoXoaLO55wDW-QPubrQCdjAB6Kn0o5ZTaQrTj4/edit#slide=id.g10316dcac78_0_6), with validation.
- Adds custom made folder to be used for testing purposes
- Moves loaders one folder up and changes predefined index into finding the index from the header
- Updates avi2mp4 script to include an argument for the folder where to start the search.

Misc:
- Updates geppetto-meta version to the latest hotfix

The ingest all script starts by cleaning a specified output directory, then parses and validates data from the two apps (NeuroScan and PromoterDB). Any issues encountered during parsing are logged to a file. If no critical errors are found, the script exports the validated data to an output directory in a structured format (CSV and JSON). Optionally, if the --transform flag is set, the script can zip CPHATE plot data and convert AVI files to MP4 format. If the --dry-run flag is not set, the script proceeds to ingest this exported data into a database.

So far I have only tested the general flow of the script with a very small custom made dataset:
- I cleared a database, keeping only the tables
- Ran the ingestion and verified that the csvs and json were created as well as the issues.log and that the entities were successfully ingested in the database

https://github.com/MetaCell/geppetto-NeuroSCAN/assets/19196034/e3c267ea-0f2e-4097-a9ba-1089896b6cef

I couldn't test with client's dataset because it doesn't follow the rules defined. cc @ddelpiano 
Here are a couple of issues reported:
- NeuroScan application name should be lowercase: neuroscan
- There are files inside the neurons folder named SVV_ColorLegend.lgd which the purpose is unclear
- Synapses are missing the pre and post synapses folder (as I see it, their format makes more sense but I followed what was stipulated in the data model)
- A lot of synapses don't follow the agreed named convention: missing zs information, using different abbreviations for the position section (f.e. post vs postSyn), destination neurons using _ to concatenate instead of & , using hyphens instead of tilde for the section part.
- Missing cphate.xls, contacts.xls and promoters.xls
- CSV's with slightly different headers (Iter1 vs iter1)

In the process of doing this task I also noticed a couple of potential problems / missing components in our data model.
- What does promoter's otherCells in strapi model map to in the promoters csv?
- There seems to be no mapping between columns Strain information and Detailed expression patterns to the promoter model we have in strapi
- The UID attribute in  neurons and contacts is not a Unique Identifier but rather just a name. An unique identifier would need to include the timepoint. No problems arise from this, it's just misnamed / confusing

The script might still need some adjustments to be completely equivalent to how it was previously used.
I'm running it locally with:

python ingestall.py --root-dir ./data_new --transform --dry-run

which outputs the csv's to be ingested in geppetto-NeuroSCAN/ingestion/output 

EDIT:

Added instructions on how to run the validator:
image

### CODE REVIEW

**Comment by afonsobspinto** [comment]
Moving the PR to draft as we need to use the previous ingestion process as the source of truth instead of the yale model presentation

**Comment by afonsobspinto** [comment]
I compared the parsers output from both old and new ingestion and they match for most of it.

There are two meaningful differences:

- neurons added in the new ingestion compared to the previous one:
```
BWM-DL01
BWM-DL02
BWM-DL03
BWM-DL04
BWM-DL05
BWM-DL06
BWM-DL08
BWM-DR01
BWM-DR02
BWM-DR03
BWM-DR04
BWM-DR05
BWM-DR06
BWM-DR07
BWM-DR08
BWM-VL01
BWM-VL02
BWM-VL03
BWM-VL04
BWM-VL05
BWM-VL06
BWM-VL07
BWM-VL08
BWM-VR01
BWM-VR02
BWM-VR03
BWM-VR04
BWM-VR05
BWM-VR06
BWM-VR07
BWM-VR08
```

- synapses not added in the new ingestion (due to invalid neuron references, aka at least one of the neurons mentioned in the synapse does not exist):
```
SVV_ADLRundefinedAIAR_CEPshVR-A_post1	
SVV_ADLRundefinedAIAR_CEPshVR-A_post2	
SVV_ADLRundefinedAIAR_CEPshVR-A_pre
SVV_AIMLundefinedASJL_CEPshVL-A_post1	
SVV_AIMLundefinedASJL_CEPshVL-A_post2	
SVV_AIMLundefinedASJL_CEPshVL-A_pre
SVV_AINLundefinedCEPshDR-A_post1	
SVV_AINLundefinedCEPshDR-A_pre
SVV_AINLundefinedCEPshVL-A_post1	
SVV_AINLundefinedCEPshVL-A_pre	
SVV_AINLundefinedCEPshVR_BAGL-A_post1	
SVV_AINLundefinedCEPshVR_BAGL-A_post2	
SVV_AINLundefinedCEPshVR_BAGL-A_pre
SVV_AINRundefinedCEPVL_CEPshVL-A_post1	
SVV_AINRundefinedCEPVL_CEPshVL-A_post2	
SVV_AINRundefinedCEPVL_CEPshVL-A_pre
SVV_ASIRundefinedAWCR_CEPshVR-A_post1	
SVV_ASIRundefinedAWCR_CEPshVR-A_post2	
SVV_ASIRundefinedAWCR_CEPshVR-A_pre	
SVV_ASJLundefinedPVQL_CEPshDL-A_post1	
SVV_ASJLundefinedPVQL_CEPshDL-A_post2	
SVV_ASJLundefinedPVQL_CEPshDL-A_pre
SVV_BDULundefinedBDUR_GLRDR-A_post1	
SVV_BDULundefinedBDUR_GLRDR-A_post2	
SVV_BDULundefinedBDUR_GLRDR-A_pre
SVV_DVCundefinedCEPshVL_RIBL-A_post1	
SVV_DVCundefinedCEPshVL_RIBL-A_post2	
SVV_DVCundefinedCEPshVL_RIBL-A_pre
SVV_IL2DLundefinedGLRDL-A_post1	
SVV_IL2DLundefinedGLRDL-A_pre
SVV_PVQRundefinedAIAR_CEPshVR-A_post1	
SVV_PVQRundefinedAIAR_CEPshVR-A_post2	
SVV_PVQRundefinedAIAR_CEPshVR-A_pre
SVV_RIARundefinedCEPVL_GLRVL-A_post1	
SVV_RIARundefinedCEPVL_GLRVL-A_post2	
SVV_RIARundefinedCEPVL_GLRVL-A_pre
SVV_RIBRundefinedAVAR_RIAR_BAGL_CEPshVR-A_post1	
SVV_RIBRundefinedAVAR_RIAR_BAGL_CEPshVR-A_post2	
SVV_RIBRundefinedAVAR_RIAR_BAGL_CEPshVR-A_post3	
SVV_RIBRundefinedAVAR_RIAR_BAGL_CEPshVR-A_post4	
SVV_RIBRundefinedAVAR_RIAR_BAGL_CEPshVR-A_pre
SVV_RIFLundefinedAVEL_CEPshVL-A_post1	
SVV_RIFLundefinedAVEL_CEPshVL-A_post2	
SVV_RIFLundefinedAVEL_CEPshVL-A_pre
SVV_RIGLundefinedCEPshVL_RIR-A_post1	
SVV_RIGLundefinedCEPshVL_RIR-A_post2	
SVV_RIGLundefinedCEPshVL_RIR-A_pre
SVV_RIHundefinedIL2L_GLRL-A_post1	
SVV_RIHundefinedIL2L_GLRL-A_post2	
SVV_RIHundefinedIL2L_GLRL-A_pre	
SVV_RIHundefinedIL2L_GLRL-B_post1	
SVV_RIHundefinedIL2L_GLRL-B_post2	
SVV_RIHundefinedIL2L_GLRL-B_pre
SVV_RIHundefinedRIAL_RIBL_CEPshVL-A_post1	
SVV_RIHundefinedRIAL_RIBL_CEPshVL-A_post2	
SVV_RIHundefinedRIAL_RIBL_CEPshVL-A_post3	
SVV_RIHundefinedRIAL_RIBL_CEPshVL-A_pre
SVV_RIHundefinedRIBL_CEPshVL-A_post1	
SVV_RIHundefinedRIBL_CEPshVL-A_post2	
SVV_RIHundefinedRIBL_CEPshVL-A_pre
SVV_RIVRundefinedGLRVL_excgl-A_post1	
SVV_RIVRundefinedGLRVL_excgl-A_post2	
SVV_RIVRundefinedGLRVL_excgl-A_pre	
SVV_RIVRundefinedGLRVL_GLRVR-A_post1	
SVV_RIVRundefinedGLRVL_GLRVR-A_post2	
SVV_RIVRundefinedGLRVL_GLRVR-A_pre	
SVV_RIVRundefinedGLRVL_SAADL-A_post1	
SVV_RIVRundefinedGLRVL_SAADL-A_post2	
SVV_RIVRundefinedGLRVL_SAADL-A_pre
SVV_SMBVLundefinedBWM-VL04_BWM-VL02_GLRL-A_post1	
SVV_SMBVLundefinedBWM-VL04_BWM-VL02_GLRL-A_post2	
SVV_SMBVLundefinedBWM-VL04_BWM-VL02_GLRL-A_post3	
SVV_SMBVLundefinedBWM-VL04_BWM-VL02_GLRL-A_pre	
SVV_SMBVLundefinedBWM-VL04_BWM-VL02_GLRL-B_post1	
SVV_SMBVLundefinedBWM-VL04_BWM-VL02_GLRL-B_post2	
SVV_SMBVLundefinedBWM-VL04_BWM-VL02_GLRL-B_post3	
SVV_SMBVLundefinedBWM-VL04_BWM-VL02_GLRL-B_pre
```
Both differences seem to me, correct in the newer version but please let me know.

Here are the files produced
[parsers_output.tar.gz](https://github.com/MetaCell/geppetto-NeuroSCAN/files/12560999/parsers_output.tar.gz)

**Comment by mohler** [comment]
Alphonso-

Synapses marked to include non-neuronal cells are not necessarily wrong.
Some synapses appear to involve other cell types.  The annotations
correctly reflect the observation of the annotator.  PLEASE, do not trash
them.  Accommodate them.

Thanks,
Bill Mohler

**Comment by afonsobspinto** [comment]
Hi Bill, no worries about the name mix-up.

Your input was duly noted. We'll discuss this internally to ensure we're aligned and update this thread after.

Thanks for reaching out

**Comment by mohler** [comment]
Great!  Thanks for the reply.

**Comment by afonsobspinto** [comment]
For now, the validator was adapted to mark the unrecognized references to a neuron in a synapse name as a warning instead of an error. 
This makes the ingestion step still process that entry, only creating relations (with neurons) for the ones it can find:

image
(SVV_ADLRundefinedAIAR_CEPshVR-A_post1, synapse previously mention as left out but now ingested)

**Comment by mohler** [comment]
That's great.  Thank you.

**Review by ddelpiano — APPROVED** [review]

**Comment by ddelpiano** [comment]
hi @mohler ,
as discussed with Noelle at our biweekly meetings we are going to accommodate the change request made and ingest these synapses, in the long term the search mechanism will be modified to make these data 'searchable' since the data model and component originally design following the requirements given don't allow the user to search for this data.
I will keep you posted on future updates related to these data.

Thanks,
Dario

**Comment by mohler** [comment]
Thank you!

**Comment by mohler** [comment]
Thank you!!

File: ingestion/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+1. Install the necessary packages:
+
+ ```
+pip install -r requirements.txt
+```
+
+2. Make sure python knows where to look up for the modules:
+```
+export PYTHONPATH=$PYTHONPATH:/path/to/geppetto-NeuroSCAN
```
Comment (zsinnema) [inline ingestion/README.md:46]: suggestion:: wouldn't it be nicer to add the geppetto-NeuroSCAN package to the python path using `pip install`? @afonsobspinto

**Review by zsinnema — CHANGES_REQUESTED** [review]

File: ingestion/README.md (hunk 1 above)
Comment (afonsobspinto) [reply ingestion/README.md:46]: :+1:

**Review by zsinnema — APPROVED** [review]

**Pull request merged**